### ISSUE TITLE
Mojom Compiler: Be more consistent about where type-related reserved words are allowed.

### ISSUE THREAD

**Issue by MitchRudominer** [domokit/mojo#622]
The following is a list of type-related reserved words that the parser treats specially in some contexts but not others:
```
array bool double float handle int8 int16 int64 map string uint8 uint16 uint32 uint64
```
In contexts where the parser is expecting a type reference these words have a special meaning and so a user may not use them as part of an identifier for a user-defined type.

But in other contexts, where the parser is not expecting a type reference, the parser currently allows these words as identifiers. Thus the parser allows the following:
```
const int32 array = 42;
struct MyStruct {
  int32 foo = array;
};
```
and in fact this generates code that compiles just fine.
But that is because `array` is not a reserved word in our target languages. On the other hand if we use `bool` instead of `array` as in
```
const int32 bool = 42;
struct MyStruct {
  int32 foo = bool;
};
```
then the parser also allows this but it generates code the fails to compile.
The Mojom spec currently calls all of these words reserved words and claims that they may not be used at all in identifiers. We need to make the parser consistent with itself and consistent with the spec. The easiest solution is to abide by the current language of the spec and forbid use of these words as user-defined identifiers in all contexts.

**Comment by viettrungluu-cr** [comment]
It seems a bit draconian to do so. Having context-dependent names seems
better to me, especially if we ever want to add basic "types" and such.

(The fact that they're mostly global keywords was not a considered choice,
but just the way the lexer was originally written.)

**Comment by MitchRudominer** [comment]
As a point of comparison, the Java language has a list of globally reserved words that may not be used in identifiers:
http://docs.oracle.com/javase/tutorial/java/nutsandbolts/_keywords.html

**Comment by MitchRudominer** [comment]
I agree that the Draconian solution of forbidding everything everywhere is not the only solution. The purpose of this bug is to track the fact that:
(1) The behavior of the parser is not exactly what is described in the spec
(2) The parser accepts some things that the backend compilers reject which is confusing.
(3) The parser's current behavior is partially an accident of implementation rather than by design.

**Comment by viettrungluu-cr** [comment]
FWIW, "draconian" wasn't meant to be an accusation. :)

I agree that some keywords are desirable/necessary, if only to make parsing
easier/feasible. Examples of keywords of this class include "struct",
"interface", and "const".

However, things like "int8" aren't really required -- wherever you'd accept
"int8", you would also have accepted "foobar" (or "int128"), based on the
possibility that there's a struct or interface with that name (and only
determine later that it's invalid).

"array"/"map"/"handle" are a little odd, but perhaps it'd be OK to
generically accept "somename<...>" at the basic parsing stage (though what
should be accepted for "..." is a good question -- so maybe just
special-casing "array"/"map"/"handle" would be better for now).

The principle here is that if we were to add an "int128" type, or a
"set<...>" 'generic', doing so should invalidate a minimal amount of
existing mojom.

For your breakdown: We should resolve (1) one way or another (or in a
different way). (2) is definitely a bug, the parser and/or backends
(realistically, the backends have many, many problems with respect to
names, keywords, etc.). And for (3), we should consider what we want the
mojom language to be like (and that's mostly what my comments are about).

**Comment by j9brown** [comment]
It sounds like the problem mainly has to do with identifier munging rather
than any kind of parser ambiguity.

Usually you end up wanting reserved words because the compiler can't parse
it otherwise.

Fwiw, I am fond of the C# convention of allowing identifiers which collide
with reserved words to be escaped using the @ prefix.

We could adopt a similar convention for mojom, if desired.  eg.  I could
declare a field of type bool called "bool" like so:

bool @bool;

There will still remain the problem of having to mangle these names to
satisfy rules in the target language but that problem won't go away no
matter what we do, given that the set of targets is potentially unbounded.
